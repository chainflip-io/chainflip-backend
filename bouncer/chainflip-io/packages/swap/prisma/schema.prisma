// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  channelId             BigInt
  srcChain              Chain
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@unique([issuedBlock, srcChain, channelId])
  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal             @db.Decimal(30, 0)
  intermediateAmount        Decimal?            @db.Decimal(30, 0)
  egressAmount              Decimal?            @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id                  BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId            BigInt
  chain               Chain
  amount              Decimal    @db.Decimal(30, 0)
  scheduledAt         DateTime
  scheduledBlockIndex String
  swap                Swap?      @relation(fields: [swapId], references: [id])
  swapId              BigInt?    @unique
  broadcast           Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId         BigInt?
  createdAt           DateTime   @default(now())
  updatedAt           DateTime   @default(now()) @updatedAt

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id                  BigInt    @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId            BigInt
  chain               Chain
  egresses            Egress[]
  requestedAt         DateTime
  requestedBlockIndex String
  succeededAt         DateTime?
  succeededBlockIndex String?
  abortedAt           DateTime?
  abortedBlockIndex   String?
  createdAt           DateTime  @default(now())
  updatedAt           DateTime  @default(now()) @updatedAt

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
