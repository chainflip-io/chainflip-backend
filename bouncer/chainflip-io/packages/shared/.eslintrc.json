{
  "extends": "../../.eslintrc.json",
  "plugins": ["eslint-plugin-n"],
  "rules": {
    "n/no-process-env": ["error"]
  },
  "overrides": [
    {
      "files": ["*.test.ts", "*.mjs", "*.js", "*.cjs"],
      "rules": {
        "n/no-process-env": ["off"]
      }
    }
  ]
}
