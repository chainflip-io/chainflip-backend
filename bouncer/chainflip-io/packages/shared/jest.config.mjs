// eslint-disable-next-line import/no-relative-packages
import baseConfig from '../../jest.config.mjs';

process.env.RPC_BROKER_HTTPS_URL = 'wss://rpc-broker.example.com';

export default {
  ...baseConfig,
};
