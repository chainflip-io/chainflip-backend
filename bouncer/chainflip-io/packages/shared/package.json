{
  "name": "@chainflip-io/sdk-shared",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "typecheck": "tsc --noEmit",
    "clean": "rm -rf dist",
    "build": "tsc -p tsconfig.cjs.json && tsc -p tsconfig.mjs.json",
    "build:abis": "typechain --target=ethers-v5 --out-dir=./src/abis ./abis/**",
    "lint": "pnpm eslint --max-warnings 0 './**/*.ts'",
    "test": "jest",
    "test:ci": "CI=1 pnpm test -- --forceExit"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "zod": "^3.21.4"
  },
  "dependencies": {
    "@polkadot/util": "^12.0.1",
    "@polkadot/util-crypto": "^12.0.1",
    "ethers": "5.7.2",
    "rxjs": "^7.8.0",
    "ws": "^8.13.0"
  },
  "devDependencies": {
    "@ethersproject/abi": "^5.7.0",
    "@ethersproject/providers": "^5.7.2",
    "@typechain/ethers-v5": "^11.0.0",
    "@types/ws": "^8.5.4",
    "typechain": "^8.2.0",
    "winston": "^3.8.2",
    "zod": "^3.21.4"
  }
}
