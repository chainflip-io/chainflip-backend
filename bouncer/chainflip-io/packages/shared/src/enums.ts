type ArrayToMap<T extends readonly string[]> = {
  [K in T[number]]: K;
};

const arrayToMap = <const T extends readonly string[]>(
  array: T,
): ArrayToMap<T> =>
  Object.fromEntries(array.map((key) => [key, key])) as ArrayToMap<T>;

export const Chains = arrayToMap(['Bitcoin', 'Ethereum', 'Polkadot', 'Arbitrum']);
export type Chain = (typeof Chains)[keyof typeof Chains];

export const Assets = arrayToMap(['FLIP', 'USDC', 'DOT', 'ETH', 'BTC', 'ARBETH', 'ARBUSDC']);
export type Asset = (typeof Assets)[keyof typeof Assets];

export const ChainflipNetworks = arrayToMap([
  'sisyphos',
  'perseverance',
  'mainnet',
  'partnernet',
]);

export type ChainflipNetwork =
  (typeof ChainflipNetworks)[keyof typeof ChainflipNetworks];

export const isTestnet = (network: ChainflipNetwork): boolean =>
  network !== ChainflipNetworks.mainnet;

export const assetChains = {
  [Assets.ETH]: Chains.Ethereum,
  [Assets.FLIP]: Chains.Ethereum,
  [Assets.USDC]: Chains.Ethereum,
  [Assets.BTC]: Chains.Bitcoin,
  [Assets.DOT]: Chains.Polkadot,
  [Assets.ARBETH]: Chains.Arbitrum,
  [Assets.ARBUSDC]: Chains.Arbitrum,
} satisfies Record<Asset, Chain>;

export const assetDecimals = {
  [Assets.DOT]: 10,
  [Assets.ETH]: 18,
  [Assets.FLIP]: 18,
  [Assets.USDC]: 6,
  [Assets.BTC]: 8,
  [Assets.ARBETH]: 18,
  [Assets.ARBUSDC]: 6,
} satisfies Record<Asset, number>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains/assets.rs#L51-L59
export const assetContractIds: Record<Asset, number> = {
  // 0 is reservered for particular cross chain messaging scenarios where we want to pass
  // through a message without making a swap.
  [Assets.ETH]: 1,
  [Assets.FLIP]: 2,
  [Assets.USDC]: 3,
  [Assets.DOT]: 4,
  [Assets.BTC]: 5,
  [Assets.ARBETH]: 6,
  [Assets.ARBUSDC]: 7,
};

export const chainAssets = {
  [Chains.Ethereum]: [Assets.ETH, Assets.USDC, Assets.FLIP],
  [Chains.Bitcoin]: [Assets.BTC],
  [Chains.Polkadot]: [Assets.DOT],
  [Chains.Arbitrum]: [Assets.ARBETH, Assets.ARBUSDC],
} satisfies Record<Chain, Asset[]>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains.rs#L52-L56
export const chainContractIds: Record<Chain, number> = {
  [Chains.Ethereum]: 1,
  [Chains.Polkadot]: 2,
  [Chains.Bitcoin]: 3,
  [Chains.Arbitrum]: 4,
};
