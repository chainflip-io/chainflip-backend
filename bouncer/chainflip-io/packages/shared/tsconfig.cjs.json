{
  "extends": "./tsconfig",
  "compilerOptions": {
    "module": "CommonJS",
    "outDir": "./dist/cjs" /* Redirect output structure to the directory. */,
    "noEmit": false
  }
}
