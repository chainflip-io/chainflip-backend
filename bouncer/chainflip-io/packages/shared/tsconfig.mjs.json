{
  "extends": "./tsconfig",
  "compilerOptions": {
    "module": "ESNext",
    "outDir": "./dist/mjs" /* Redirect output structure to the directory. */,
    "noEmit": false
  }
}
