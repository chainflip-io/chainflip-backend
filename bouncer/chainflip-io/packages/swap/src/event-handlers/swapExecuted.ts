import { z } from 'zod';
import { u128, u64 } from '@/shared/parsers';
import type { EventHandlerArgs } from '.';

const swapExecutedArgs = z.object({
  swapId: u64,
  egressAmount: u128,
  intermediateAmount: u128.optional(),
});

export type SwapExecutedEvent = z.input<typeof swapExecutedArgs>;

export default async function swapExecuted({
  prisma,
  block,
  event,
}: EventHandlerArgs): Promise<void> {
  const { swapId, egressAmount, intermediateAmount } = swapExecutedArgs.parse(
    event.args,
  );

  // use updateMany to skip update if we are not tracking swap
  await prisma.swap.updateMany({
    where: { nativeId: swapId },
    data: {
      egressAmount: egressAmount.toString(),
      intermediateAmount: intermediateAmount?.toString(),
      swapExecutedAt: new Date(block.timestamp),
      swapExecutedBlockIndex: `${block.height}-${event.indexInBlock}`,
    },
  });
}
