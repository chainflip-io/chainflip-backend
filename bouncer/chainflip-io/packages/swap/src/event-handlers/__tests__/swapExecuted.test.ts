import { Assets } from '@/shared/enums';
import prisma from '../../client';
import swapExecuted from '../swapExecuted';
import {
  DOT_ADDRESS,
  buildSwapExecutedMock,
  createDepositChannel,
} from './utils';

describe(swapExecuted, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap" CASCADE`;
  });

  it('updates an existing swap with the execution timestamp', async () => {
    const {
      eventContext: { event },
      block,
    } = buildSwapExecutedMock({
      swapId: '9876545',
      egressAmount: '10000000000',
    });

    const { swapId } = event.args;

    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createDepositChannel({
      swaps: {
        create: {
          nativeId: BigInt(swapId),
          depositAmount: '10000000000',
          depositReceivedAt: new Date(block.timestamp - 6000),
          depositReceivedBlockIndex: `${block.height}-${event.indexInBlock}`,
          srcAsset: Assets.ETH,
          destAsset: Assets.DOT,
          destAddress: DOT_ADDRESS,
        },
      },
    });

    await prisma.$transaction((tx) =>
      swapExecuted({
        block: block as any,
        event: event as any,
        prisma: tx,
      }),
    );

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
    });

    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
    });
  });

  it('updates an existing swap with the execution timestamp and intermediate amount', async () => {
    const {
      eventContext: { event },
      block,
    } = buildSwapExecutedMock({
      swapId: '9876545',
      egressAmount: '10000000000',
      intermediateAmount: '100000',
    });

    const { swapId } = event.args;

    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createDepositChannel({
      swaps: {
        create: {
          nativeId: BigInt(swapId),
          depositAmount: '10000000000',
          depositReceivedAt: new Date(block.timestamp - 6000),
          depositReceivedBlockIndex: `${block.height}-${event.indexInBlock}`,
          srcAsset: Assets.ETH,
          destAsset: Assets.DOT,
          destAddress: DOT_ADDRESS,
        },
      },
    });

    await prisma.$transaction((tx) =>
      swapExecuted({
        block: block as any,
        event: event as any,
        prisma: tx,
      }),
    );

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
    });

    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
    });
  });
});
