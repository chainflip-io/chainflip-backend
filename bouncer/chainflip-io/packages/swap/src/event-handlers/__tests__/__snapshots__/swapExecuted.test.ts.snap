// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swapExecuted updates an existing swap with the execution timestamp 1`] = `
{
  "createdAt": Any<Date>,
  "depositAmount": "10000000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "100-0",
  "destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX",
  "destAsset": "DOT",
  "egressAmount": "10000000000",
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 9876545n,
  "srcAsset": "ETH",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": 2022-12-06T14:31:39.000Z,
  "swapExecutedBlockIndex": "100-0",
  "txHash": null,
  "updatedAt": Any<Date>,
}
`;

exports[`swapExecuted updates an existing swap with the execution timestamp and intermediate amount 1`] = `
{
  "createdAt": Any<Date>,
  "depositAmount": "10000000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "100-0",
  "destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX",
  "destAsset": "DOT",
  "egressAmount": "10000000000",
  "id": Any<BigInt>,
  "intermediateAmount": "100000",
  "nativeId": 9876545n,
  "srcAsset": "ETH",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": 2022-12-06T14:31:39.000Z,
  "swapExecutedBlockIndex": "100-0",
  "txHash": null,
  "updatedAt": Any<Date>,
}
`;
